File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        def count_sort(word: str) -> str:
            counts = [0] * 26
            for char in word:
                counts[ord(char) - ord('a')] += 1

            result = []
            for index, count in enumerate(counts):
                result.append(chr(index + ord('a'))*count)
            return ''.join(result)            
        
        hash_map = collections.defaultdict(list)

        for word in strs:
            hash_map[count_sort(word)].append(word)
        
        return list(hash_map.values())
            

File: 0049-group-anagrams/0049-group-anagrams.go
func countSort(word string) string {
        counts := make([]int, 26)
        for _, char := range word {
            counts[int(char) - int('a')]++
        }
        var builder strings.Builder
        for index, count := range counts {
            for i := 0; i < count; i++ {
                builder.WriteByte(byte(index + int('a')))
            }
        }
        return builder.String()
}

func groupAnagrams(strs []string) [][]string {

    hashMap := make(map[string][]string)

    for _, word := range strs {
        sortedWord := countSort(word)
        if _, ok := hashMap[sortedWord]; !ok {
            hashMap[sortedWord] = make([]string, 0)
        }
        hashMap[sortedWord] = append(hashMap[sortedWord], word)
    }

    result := [][]string{}
    for _, value := range hashMap {
        result = append(result, value)
    }
    
    return result
}